aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,319 @@
+    BLS12381G2AffineCompressed,
+    BLS12381G2AffineUnompressed,
+    BLS12381Gt,
+    BLS12381FrLsb,
+    BLS12381FrMsb,
+}
+
+impl TryFrom<u64> for SerializationFormat {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        match value {
+            0x0400000000000000 => Ok(SerializationFormat::BLS12381Fq12LscLscLscLsb),
+            0x0600000000000000 => Ok(SerializationFormat::BLS12381G1AffineUncompressed),
+            0x0601000000000000 => Ok(SerializationFormat::BLS12381G1AffineCompressed),
+            0x0800000000000000 => Ok(SerializationFormat::BLS12381G2AffineUnompressed),
+            0x0801000000000000 => Ok(SerializationFormat::BLS12381G2AffineCompressed),
+            0x0900000000000000 => Ok(SerializationFormat::BLS12381Gt),
+            0x0A00000000000000 => Ok(SerializationFormat::BLS12381FrLsb),
+            0x0A01000000000000 => Ok(SerializationFormat::BLS12381FrMsb),
+            _ => Err(()),
+        }
+    }
+}
+
+/// Hash-to-structure suites.
+pub enum HashToStructureSuite {
+    Bls12381g1XmdSha256SswuRo,
+    Bls12381g2XmdSha256SswuRo,
+}
+
+impl TryFrom<u64> for HashToStructureSuite {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        match value {
+            0x0001000000000000 => Ok(HashToStructureSuite::Bls12381g1XmdSha256SswuRo),
+            0x0002000000000000 => Ok(HashToStructureSuite::Bls12381g2XmdSha256SswuRo),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Tid, Default)]
+pub struct AlgebraContext {
+    objs: Vec<Rc<dyn Any>>,
+}
+
+impl AlgebraContext {
+    pub fn new() -> Self {
+        Self { objs: Vec::new() }
+    }
+}
+
+macro_rules! structure_from_ty_arg {
+    ($context:expr, $typ:expr) => {{
+        let type_tag = $context.type_to_type_tag($typ).unwrap();
+        Structure::from_type_tag(&type_tag)
+    }};
+}
+
+macro_rules! store_obj {
+    ($context:expr, $obj:expr) => {{
+        let target_vec = &mut $context.extensions_mut().get_mut::<AlgebraContext>().objs;
+        let ret = target_vec.len();
+        target_vec.push(Rc::new($obj));
+        ret
+    }};
+}
+
+macro_rules! abort_unless_feature_enabled {
+    ($context:ident, $feature:expr) => {
+        if !$context.get_feature_flags().is_enabled($feature) {
+            return Err(SafeNativeError::Abort {
+                abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+            });
+        }
+    };
+}
+
+macro_rules! get_obj_pointer {
+    ($context:expr, $handle:expr) => {{
+        $context.extensions_mut().get_mut::<AlgebraContext>().objs[$handle].clone()
+    }};
+}
+
+macro_rules! ark_serialize_internal {
+    (
+        $gas_params:expr,
+        $context:expr,
+        $structure:expr,
+        $handle:expr,
+        $format:expr,
+        $typ:ty,
+        $ser_func:ident
+    ) => {{
+        let element_ptr = get_obj_pointer!($context, $handle);
+        let element = element_ptr.downcast_ref::<$typ>().unwrap();
```

Comment:
If unwrap() fails here... we are toast... we need to handle this properly & abort.

Reply:
I'd say here the `downcast_ref` failing is something unreachable, guaranteed by the implementation.
Do we really need to handle?

Reply:
@vgao1996 @wrwg what would you say?

Reply:
Good question. I'd say that in an ideal world, `unreachable` or `unwrap` are conceptually sound, provided that you have adequate testing for all of your components. However we all know we are not in that world yet, so I'd recommend you return an `Err(SafeNativeError::InvariantViolation((PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))))`

Reply:
Fixed.

Reply:
Indeed, we need to be more defensive when writing natives. Sure, the invariants are supposed to guarantee things, but sometimes invariants don't hold due to bugs. Best to not mess around in a native.